app/models.py

```diff
@@ -20,14 +22,112 @@
+        parameter is set to False.
+    use_auto_verify: boolean
+        if True (default), will call verify and commit
+        functions of the session before exiting the context.
+        If True, must specify a legal User object.
+
+    Returns
+    -------
+    a scoped session object that can be used in a context
+    manager to access the database. If auto verify is enabled,
+    will use the current user given to apply verification
+    and commit to the database when exiting context.
+
+    """
+
+    def verify(self):
+        # import here to avoid circular import loop
+        from .handlers.base import bulk_verify
+
+        """Check that the current user has permission to create, read,
+        update, or delete rows that are present in the session. If not,
+        raise an AccessError (causing the transaction to fail and the API to
+        respond with 401).
+
+        """
+        # get items to be inserted
+        new_rows = [row for row in self.new]
+
+        # get items to be updated
+        updated_rows = [row for row in self.dirty if self.is_modified(row)]
+
+        # get items to be deleted
+        deleted_rows = [row for row in self.deleted]
+
+        # get items that were read
+        read_rows = [
+            row
+            for row in set(self.identity_map.values())
+            - (set(updated_rows) | set(new_rows) | set(deleted_rows))
+        ]
+
+        # need to check delete permissions before flushing, as deleted records
+        # are not present in the transaction after flush (thus can't be used in
+        # joins). Read permissions can be checked here or below as they do not
+        # change on flush.
+        for mode, collection in zip(
+            ["read", "update", "delete"],
+            [read_rows, updated_rows, deleted_rows],
+        ):
+            bulk_verify(mode, collection, self.user_or_token)
+
+        # update transaction state in DB, but don't commit yet. this updates
+        # or adds rows in the database and uses their new state in joins,
+        # for permissions checking purposes.
+        self.flush()
+        bulk_verify("create", new_rows, self.user_or_token)
+
+    if use_auto_verify and user_or_token is None:
+        raise RuntimeError(
+            "Cannot start a session with use_auto_verify without a valid user."
+        )
+    if user_or_token is not None and not isinstance(user_or_token, (User, Token)):
+        raise ValueError(
+            "user_or_token must be an instance of User or Token, "
+            f"got {user_or_token.__class__.__name__}."
+        )
+    with scoped_session(sessionmaker(), scopefunc=session_context_id.get)() as session:
+        session.user_or_token = user_or_token
+        session.use_auto_verify = use_auto_verify
+        # make this an instance method of session
+        session.verify = types.MethodType(verify, session)
+        yield session
+
+        session.expunge_all()  # make sure objects persist after session closes
```

Comment:
Why is this necessary? Do we want objects to persist after the session closes?

Reply:
Yes. When committing the session, objects we got from it become stale. There are lots of cases where you keep using them, even when they are detached from the DB. Example: use a `with Session() as session` inside a function that returns an object you need. The function quits and shuts down the session but the object is still usable.

Reply:
If we do not commit the session we probably don't need the expunge anymore. But is it ok to keep using an object loaded after a session that was closed (but not committed)?

Reply:
I don't think you can use objects retrieved outside of the session they came from.

Reply:
Yeah, you must. For example the `get_if_accessible_by` function returns an object. There's no way to recreate this pattern if you create a session and want to pass the object out in the return statement. 
Besides, every time a handler gives back data, the session ends and the data persists. 

I think the only thing that is not allowed is to modify the objects after the session is closed, which makes sense.

Reply:
That's correct—no writing to those objects.

Reply:
OK. I've also removed the `expunge_all()` now that we are not committing automatically. 
That means that commit is only used when making changes to the DB. 
When only getting information, we must rely on the select statements that go through baselayer to give us only accessible rows, and not the `verify()` command.